certificateutil/filter.go

```diff
@@ -22,26 +22,36 @@ type ValidCertificateInfo struct {
 
 // FilterValidCertificateInfos filters out invalid and duplicated common name certificaates
 func FilterValidCertificateInfos(certificateInfos []CertificateInfoModel) ValidCertificateInfo {
-	certificateInfosByName := map[string]CertificateInfoModel{}
-
-	var invalidCertificates, duplicatedCertificates []CertificateInfoModel
+	var invalidCertificates []CertificateInfoModel
+	nameToCerts := map[string][]CertificateInfoModel{}
 	for _, certificateInfo := range certificateInfos {
 		if certificateInfo.CheckValidity() != nil {
 			invalidCertificates = append(invalidCertificates, certificateInfo)
 			continue
 		}
-		activeCertificate, ok := certificateInfosByName[certificateInfo.CommonName]
-		if !ok {
-			certificateInfosByName[certificateInfo.CommonName] = certificateInfo
-		} else if certificateInfo.EndDate.After(activeCertificate.EndDate) {
-			duplicatedCertificates = append(duplicatedCertificates, activeCertificate)
-			certificateInfosByName[certificateInfo.CommonName] = certificateInfo
-		}
+
+		nameToCerts[certificateInfo.CommonName] = append(nameToCerts[certificateInfo.CommonName], certificateInfo)
 	}
 
-	validCertificates := []CertificateInfoModel{}
-	for _, validCertificate := range certificateInfosByName {
-		validCertificates = append(validCertificates, validCertificate)
+	var validCertificates, duplicatedCertificates []CertificateInfoModel
+	for _, certs := range nameToCerts {
+		if len(certs) == 0 {
+			continue
+		}
+
+		latestCert := certs[0]
```

Comment:
NOTE: what about this

```
		sort.Slice(certs, func(i, j int) bool {
			return certs[i].EndDate.Before(certs[j].EndDate)
		})

		validCertificates = append(validCertificates, certs[0])
		duplicatedCertificates = append(duplicatedCertificates, certs[1:]...)
```

is not it a bit cleaner?

Reply:
It is more clear, even if less optimal in theory.